src/common/sys_lmdb.c

```diff
@@ -430,11 +427,13 @@ lmm_db_fini(void)
 	if (lmm_db.db_lock)
 		ABT_mutex_free(&lmm_db.db_lock);
 	if (lmm_db.db_file) {
-		if (lmm_db.db_env)
-			mdb_env_close(lmm_db.db_env);
 		if (lmm_db.db_destroy_db)
 			lmm_db_unlink(&lmm_db.db_pub);
-		mdb_dbi_close(lmm_db.db_env, lmm_db.db_dbi);
+		if (lmm_db.db_env) {
+			if (lmm_db.db_dbi_valid)
+				mdb_dbi_close(lmm_db.db_env, lmm_db.db_dbi);
+			mdb_env_close(lmm_db.db_env);
+		}
 		free(lmm_db.db_file);
```

Comment:
we should use D_ASPRINTF and D_FREE for these

Reply:
Sure